pkg/email/smtp.go

```diff
@@ -0,0 +1,94 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package email is logic for sending email invitations
+package email
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"mime/quotedprintable"
+	"net/smtp"
+
+	"firebase.google.com/go/auth"
+)
+
+var _ Provider = (*SMTPProvider)(nil)
+
+// SMTPProvider sends messages via an external SMTP server.
+type SMTPProvider struct {
+	FirebaseAuth *auth.Client
+
+	User     string
+	Password string
+	SMTPHost string
+	SMTPPort string
+}
+
+// NewSMTP creates a new Smtp email sender with the given auth.
+func NewSMTP(ctx context.Context, user, password, host, port string, auth *auth.Client) (Provider, error) {
+	return &SMTPProvider{
+		FirebaseAuth: auth,
+		User:         user,
+		Password:     password,
+		SMTPHost:     host,
+		SMTPPort:     port,
+	}, nil
+}
+
+// SendNewUserInvitation sends a password reset email to the user.
+func (s *SMTPProvider) SendNewUserInvitation(ctx context.Context, toEmail string) error {
+	// Header
+	header := make(map[string]string)
+	header["From"] = s.User
```

Comment:
We should try to make most of these constants. Both the map keys and many of their values are constant.